### PR TITLE
Support serializing/deserializing JsValues

### LINKED ISSUE #32: Add support for serializing and deserializing JsValue?
Let say we have a `struct` in Rust as follows:
```rust
struct Test {
   v1: JsValue,
   v2: JsValue,
}
let v1 = JsValue::from("hello");
let v2 = JsValue::from("world");
let t = Test { v1, v2 };
```
It would be nice to be able to pass `t` back and forth between Rust and Javascript, where during "serialization" `v1` and `v2` are just replaced with the actual Javascript values they represent, and on deserialization, the Javascript values are pushed onto wasm-bindgen's heap and the struct members are initialized with `JsValue`s that reference those Javascript objects on the heap.

At the moment, this doesn't work out of the box since `Serialize` is not implemented for `JsValue`, which makes sense since it's internal representation is just an index to wasm-bindgen's heap, but perhaps there is a way to hack this, possibly with `#[serde(serialize_with = "path")]` on the `JsValue` fields?

### PR DESCRIPTION
**Pull request by FredrikNoren**
Fixes #32

### CODE REVIEW

**Comment by FredrikNoren** [comment]
@RReverser thoughts? We've been using it here and seems to work fine.

**Comment by RReverser** [comment]
Have you seen #40? I believe we got a bit closer there to fix some of the potential issues with other formats, but there was some feedback remaining (it's been a while).

File: src/preserve.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+impl<'de, T: From<JsValue> + Into<JsValue> + Clone> Deserialize<'de> for PreserveJsValue<T> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct JsValueVisitor<T: From<JsValue> + Into<JsValue> + Clone>(
+            std::marker::PhantomData<T>,
+        );
```
Comment (richardpringle) [inline src/preserve.rs:54]: Is there any benefit to this type being generic? Why not just use the outer `T` that's already bound?

Comment (FredrikNoren) [reply src/preserve.rs:54]: Hm iirc you can't bind to outer generics like that in Rust? Could be wrong though

**Comment by FredrikNoren** [comment]
@RReverser Yup saw it, but it seemed like it got stuck. Anyway, feel free to close this PR if the other route is preferred, just figured I'd post it in case it would help someone else. The only drawback of this implementation that I'm aware of is that it won't work if you try to serialize from two threads simultaneously, but couldn't figure out a way to fix that.

**Comment by RReverser** [comment]
I chose to add some finishing touches to #40 myself and going to merge that one instead as it's has some robust checks against 3rd-party serializers/deserializers. Thanks for the PR anyway!

**Comment by RReverser** [comment]
I'm guessing no objections, so I'm going to publish it to Cargo.

**Comment by FredrikNoren** [comment]
@RReverser Yeah sorry I haven't had any time to test this, but I'll post an issue if I run into any problems with it when we upgrade here (but might be a little while until we have time)

**Comment by RReverser** [comment]
No problem. Thanks for your PR again.

**Pull request closed without merging**